src/Domain/FunctionsOfTime/QuaternionFunctionOfTime.cpp

```diff
@@ -0,0 +1,213 @@
+    const double t0 = omega_deriv_info[i - 1].time;
+    quaternion_to_integrate = datavector_to_quaternion(
+        stored_quaternions_and_times_[i - 1].stored_quantities[0]);
+
+    solve_quaternion_ode(make_not_null(&quaternion_to_integrate), t0, t);
+
+    // normalize quaternion
+    normalize_quaternion(make_not_null(&quaternion_to_integrate));
+
+    stored_quaternions_and_times_.emplace_back(
+        t, std::array<DataVector, 1>{
+               quaternion_to_datavector(quaternion_to_integrate)});
+  }
+
+  ASSERT(omega_deriv_info.size() == stored_quaternions_and_times_.size(),
+         "The number of stored omegas must be the same as the number of stored "
+         "quaternions after updating the missing quaternions. Now there are "
+             << omega_deriv_info.size() << " stored omegas and "
+             << stored_quaternions_and_times_.size() << " stored quaternions.");
+}
+
+template <size_t MaxDeriv>
+void QuaternionFunctionOfTime<MaxDeriv>::solve_quaternion_ode(
+    const gsl::not_null<boost::math::quaternion<double>*>
+        quaternion_to_integrate,
+    const double t0, const double t) const noexcept {
+  // lambda that stores the internals of the ode
+  const auto quaternion_ode_system =
+      [this](const boost::math::quaternion<double>& state,
+             boost::math::quaternion<double>& dt_state,
+             const double time) noexcept {
+        const boost::math::quaternion<double> omega =
+            datavector_to_quaternion(omega_f_of_t_.func(time)[0]);
+        dt_state = 0.5 * state * omega;
+      };
+
+  // Dense stepper
+  auto dense_stepper = boost::numeric::odeint::make_dense_output(
+      1.0e-15, 1.0e-14,
+      boost::numeric::odeint::runge_kutta_dopri5<
+          boost::math::quaternion<double>, double,
+          boost::math::quaternion<double>, double,
+          boost::numeric::odeint::vector_space_algebra>{});
+
+  // Integrate from t0 to t, storing result in quaternion_to_integrate
+  boost::numeric::odeint::integrate_adaptive(
+      dense_stepper, quaternion_ode_system, *quaternion_to_integrate, t0, t,
+      1e-4);
+}
+
+template <size_t MaxDeriv>
+boost::math::quaternion<double> QuaternionFunctionOfTime<MaxDeriv>::setup_func(
+    const double t) const noexcept {
+  // Get quaternion and time at closest time before t
+  const auto& stored_info_at_t0 =
+      stored_info_from_upper_bound(t, stored_quaternions_and_times_);
+  boost::math::quaternion<double> quat_to_integrate =
+      datavector_to_quaternion(stored_info_at_t0.stored_quantities[0]);
+
+  // Solve the ode and store the result in quat_to_integrate
+  solve_quaternion_ode(make_not_null(&quat_to_integrate),
+                       stored_info_at_t0.time, t);
+
+  // Make unit quaternion
+  normalize_quaternion(make_not_null(&quat_to_integrate));
+
+  return quat_to_integrate;
+}
+
+template <size_t MaxDeriv>
+std::array<DataVector, 1> QuaternionFunctionOfTime<MaxDeriv>::quat_func(
```

Comment:
Codecov points out that the functions returning fewer derivatives aren't tested.  Those are easy, so add them.

Reply:
Is there a way to run codecov locally so I don't have to wait for CI tests to run to know if I missed anything?

Reply:
I don't know.  @nilsleiffischer?

Reply:
Yes you can run it locally. Set up a separate build directory with `cmake -D COVERAGE=ON` and run `make unit-test-coverage` (maybe need to `make unit-tests`) before. You'll get the results as HTML output. That said, it might be easier to just push the code and have CI build it, since the separate build with coverage can take up a lot of space and running unit tests with coverage can take a long time (40min on CI).

Reply:
Thanks!